src/graphics/webgpu/webgpu-render-pipeline.js

```diff
@@ -0,0 +1,203 @@
+    undefined,              // BLENDMODE_CONSTANT_ALPHA
+    undefined               // BLENDMODE_ONE_MINUS_CONSTANT_ALPHA
+];
+
+// temp array to avoid allocation
+const _bindGroupLayouts = [];
+
+/**
+ * @ignore
+ */
+class WebgpuRenderPipeline {
+    constructor(device) {
+        /** @type {WebgpuGraphicsDevice} */
+        this.device = device;
+
+        /**
+         * The cache of vertex buffer layouts
+         *
+         * @type {WebgpuVertexBufferLayout}
+         */
+        this.vertexBufferLayout = new WebgpuVertexBufferLayout();
+
+        /**
+         * The cache of render pipelines
+         *
+         * @type {Map<string, object>}
+         */
+        this.cache = new Map();
+    }
+
+    // TODO: somehow pass entity name to assign to things, maybe hook up to GraphicsDebug stuff
+
+    get(primitive, vertexFormat0, vertexFormat1, shader, renderTarget, bindGroupFormats, renderState) {
+
+        // render pipeline unique key
+        const key = this.getKey(primitive, vertexFormat0, vertexFormat1, shader, renderTarget, bindGroupFormats, renderState);
+
+        // cached pipeline
+        let pipeline = this.cache.get(key);
+        if (!pipeline) {
+            console.log("create pipeline");
+
+            const primitiveTopology = _primitiveTopology[primitive.type];
+            Debug.assert(primitiveTopology, `Unsupported primitive topology ${primitive}`);
+
+            // pipeline layout
+            const pipelineLayout = this.getPipelineLayout(bindGroupFormats);
+
+            // vertex buffer layout
+            const vertexBufferLayout = this.vertexBufferLayout.get(vertexFormat0, vertexFormat1);
+
+            // pipeline
+            pipeline = this.create(primitiveTopology, shader.impl, renderTarget, pipelineLayout, renderState, vertexBufferLayout);
+            this.cache.set(key, pipeline);
+        }
+
+        return pipeline;
+    }
+
+    /**
+     * Generate a unique key for the render pipeline. Keep this function as lean as possible,
+     * as it executes for each draw call.
+     */
+    getKey(primitive, vertexFormat0, vertexFormat1, shader, renderTarget, bindGroupFormats, renderState) {
+
+        let bindGroupKey = '';
+        for (let i = 0; i < bindGroupFormats.length; i++) {
+            bindGroupKey += bindGroupFormats[i].key;
+        }
+
+        const vertexBufferLayoutKey = this.vertexBufferLayout.getKey(vertexFormat0, vertexFormat1);
+        const renderTargetKey = renderTarget.impl.key;
+        const renderStateKey = renderState.blendKey;
+
+        return vertexBufferLayoutKey + shader.impl.vertexCode + shader.impl.fragmentCode +
+            renderTargetKey + renderStateKey + primitive.type + bindGroupKey;
+    }
+
+    // TODO: this could be cached using bindGroupKey
+    getPipelineLayout(bindGroupFormats) {
+
+        bindGroupFormats.forEach((format) => {
+            _bindGroupLayouts.push(format.bindGroupLayout);
+        });
```

Comment:
Would this be better using `map`?

Reply:
This is building WebGPU descriptor, and it is an array of objects, so I'm filling it here.